### PR TITLE
PYTHON-3367 Add support for GCP attached service accounts when using GCP KMS

### PR DESCRIPTION
**Pull request by blink1073**
Integration tests are in https://github.com/mongodb/mongo-python-driver/pull/1064

### CODE REVIEW

File: bindings/python/test/test_mongocrypt.py [hunk 1]
```diff
@@ -123,16 +127,16 @@ def test_mongocrypt_options_validation(self):
                 {'aws': {'accessKeyId': 'foo'}},
                 {'aws': {'secretAccessKey': 'foo'}}]:
             with self.assertRaisesRegex(
-                    ValueError, "kms_providers\['aws'\] must contain "
+                    ValueError, r"kms_providers\['aws'\] must contain "
```
Comment (blink1073) [inline bindings/python/test/test_mongocrypt.py:130]: These changes address warnings seen while running the tests:

```python
test/test_mongocrypt.py:139
  /Users/steve.silvester/workspace/libmongocrypt/bindings/python/test/test_mongocrypt.py:139: DeprecationWarning: invalid escape sequence '\('
    "instance of bytes or str \(unicode in Python 2\)"):
```

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 2]
```diff
@@ -794,21 +800,51 @@ def __init__(self, ctx, kms_providers, filter, provider, master_key,
             raise
 
 
+def _get_gcp_credentials():
+    """Get on-demand GCP credentials"""
+    metadata_host = "metadata.google.internal"
+    if os.getenv("GCE_METADATA_HOST"):
+        metadata_host = os.environ["GCE_METADATA_HOST"]
```
Comment (ShaneHarvey) [inline bindings/python/pymongocrypt/mongocrypt.py:807]: Suggest:
```
metadata_host = os.getenv("GCE_METADATA_HOST") or "metadata.google.internal"
```

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 3]
```diff
@@ -794,21 +800,51 @@ def __init__(self, ctx, kms_providers, filter, provider, master_key,
...
+    url = "http://%s/computeMetadata/v1/instance/service-accounts/default/token" % metadata_host
+
+    headers = {"Metadata-Flavor": "Google"}
+    response = requests.get(url, headers=headers)
```
Comment (ShaneHarvey) [inline bindings/python/pymongocrypt/mongocrypt.py:811]: Do we need exception handling here?

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 4]
```diff
@@ -794,21 +800,51 @@ def __init__(self, ctx, kms_providers, filter, provider, master_key,
...
+    if response.status_code != 200:
+        msg = "Unable to retrieve GCP credentials: expected StatusCode 200, got StatusCode: %s. Response body:\n%s" % (response.status_code, response.content)
+        raise ValueError(msg)
+    try:
+        data = response.json()
+    except Exception :
```
Comment (ShaneHarvey) [inline bindings/python/pymongocrypt/mongocrypt.py:817]: `Exception :` ->  `Exception:`

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 5]
```diff
@@ -794,21 +800,51 @@ def __init__(self, ctx, kms_providers, filter, provider, master_key,
...
+    metadata_host = "metadata.google.internal"
+    if os.getenv("GCE_METADATA_HOST"):
+        metadata_host = os.environ["GCE_METADATA_HOST"]
+    url = "http://%s/computeMetadata/v1/instance/service-accounts/default/token" % metadata_host
+
+    headers = {"Metadata-Flavor": "Google"}
+    response = requests.get(url, headers=headers)
+    if response.status_code != 200:
+        msg = "Unable to retrieve GCP credentials: expected StatusCode 200, got StatusCode: %s. Response body:\n%s" % (response.status_code, response.content)
+        raise ValueError(msg)
```
Comment (ShaneHarvey) [inline bindings/python/pymongocrypt/mongocrypt.py:814]: ValueError is usually used to indicate an invalid argument was passed whereas here the problem is elsewhere. Perhaps MongoCryptError would be more appropriate here and below?

File: bindings/python/test/test_mongocrypt.py (hunk 1 above)
Comment (ShaneHarvey) [reply bindings/python/test/test_mongocrypt.py:130]: Noted.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 6]
```diff
@@ -794,21 +800,51 @@ def __init__(self, ctx, kms_providers, filter, provider, master_key,
...
     """Get on-demand kms credentials.
 
     This is a separate function so it can be overridden in unit tests."""
-    if 'aws' not in kms_providers:
-        return
-    if len(kms_providers['aws']):
+    on_demand_aws = 'aws' in kms_providers and not len(kms_providers['aws'])
+    on_demand_gcp = 'gcp' in kms_providers and not len(kms_providers['gcp'])
+    if not on_demand_aws and not on_demand_gcp:
         return
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongocrypt/mongocrypt.py:835]: Fixed

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 2 above)
Comment (blink1073) [reply bindings/python/pymongocrypt/mongocrypt.py:807]: Done

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 3 above)
Comment (blink1073) [reply bindings/python/pymongocrypt/mongocrypt.py:811]: Done

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 4 above)
Comment (blink1073) [reply bindings/python/pymongocrypt/mongocrypt.py:817]: Fixed

File: bindings/python/pymongocrypt/mongocrypt.py (hunk 5 above)
Comment (blink1073) [reply bindings/python/pymongocrypt/mongocrypt.py:814]: Done

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bindings/python/release.sh [hunk 7]
```diff
@@ -14,9 +14,9 @@ set -o xtrace   # Write all commands first to stderr
 set -o errexit  # Exit the script with error if any of the commands fail
 
 # The libmongocrypt git revision release to embed in our wheels.
-REVISION=$(git rev-list -n 1 1.5.2)
+REVISION=$(git rev-list -n 1 1.6.0)
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply bindings/python/release.sh:17]: Oh I see 1.6.1 had no code changes so 1.6.0 is fine.

File: bindings/python/pymongocrypt/mongocrypt.py [hunk 8]
```diff
@@ -794,21 +800,54 @@ def __init__(self, ctx, kms_providers, filter, provider, master_key,
...
-        )
-    creds = aws_temp_credentials()
-    creds_dict = {"accessKeyId": creds.username, "secretAccessKey": creds.password}
-    if creds.token:
-        creds_dict["sessionToken"] = creds.token
-    return { 'aws': creds_dict }
+    on_demand_aws = 'aws' in kms_providers and not len(kms_providers['aws'])
+    on_demand_gcp = 'gcp' in kms_providers and not len(kms_providers['gcp'])
+    if not on_demand_aws and not on_demand_gcp:
+        return None
```
[the remark this answers is not in the document]
Comment (blink1073) [reply bindings/python/pymongocrypt/mongocrypt.py]: Done

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**